xdp-dump/xdpdump.c

```diff
@@ -0,0 +1,628 @@
+				printf("  %s\n", hline);
+			}
+		}
+		ctx->captured_packets++;
+		break;
+
+	case PERF_RECORD_LOST:
+		ctx->missed_events += lost->lost;
+		break;
+	}
+
+	return LIBBPF_PERF_EVENT_CONT;
+}
+
+/*****************************************************************************
+ * get_epoch_to_uptime_delta()
+ *****************************************************************************/
+static uint64_t get_epoch_to_uptime_delta(void)
+{
+	/* This function will calculate the rough delta between uptime
+	 * seconds and the epoch time. This is not a precise delta as there is
+	 * a delay between calling the two functions below (and time() being in
+	 * seconds), but it's good enough to get a general offset. The delta
+	 * between packets is still based on the timestamps from the trace
+	 * infrastructure.
+	 */
+	struct timespec ts;
+	uint64_t        uptime;
+	uint64_t        epoch = time(NULL) * 1000000000ULL;
+
+	clock_gettime(CLOCK_MONOTONIC, &ts);
+	uptime = ts.tv_sec * 1000000000ULL + ts.tv_nsec;
+
+	return epoch - uptime;
+}
+
+/*****************************************************************************
+ * capture_on_legacy_interface()
+ *****************************************************************************/
+static bool capture_on_legacy_interface(struct dumpopt *cfg)
+{
+	bool              rc = false;
+	char              errbuf[PCAP_ERRBUF_SIZE];
+	uint64_t          captured_packets = 0;
+	pcap_t           *pcap = NULL;
+	pcap_dumper_t    *pcap_dumper = NULL;
+	struct pcap_stat  ps;
+
+	/* Open pcap handle for live capture. */
+	if (cfg->rx_capture != RX_FLAG_FENTRY) {
+		pr_warn("ERROR: For legacy capture only \"--rx-capture entry\""
+			" is supported!\n");
+		goto error_exit;
+	}
+
+	pcap = pcap_open_live(cfg->iface.ifname, cfg->snaplen,
+			      true, 1000, errbuf);
+	if (pcap == NULL) {
+		pr_warn("ERROR: Can't open pcap live interface: %s\n", errbuf);
+		goto error_exit;
+	}
+
+	/* Open the pcap handle for pcap file. */
+	if (cfg->pcap_file) {
+		pcap_dumper = pcap_dump_open(pcap, cfg->pcap_file);
+		if (!pcap_dumper) {
+			pr_warn("ERROR: Can't open pcap file for writing!\n");
+			goto error_exit;
+		}
+	}
+
+	/* Loop for receive packets on live interface. */
+	exit_pcap = pcap;
+	while (!exit_xdpdump) {
+		const uint8_t      *packet;
+		struct pcap_pkthdr  h;
+
+		packet = pcap_next(pcap, &h);
+		if (!packet)
+			continue;
+
+		if (pcap_dumper) {
+			pcap_dump((u_char *) pcap_dumper, &h, packet);
+		} else {
+			int i;
+			char hline[SNPRINTH_MIN_BUFFER_SIZE];
+
+			printf("%lu.%06lu: packet size %u bytes, "
+			       "captured %u bytes\n",
+			       h.ts.tv_sec, h.ts.tv_usec,
+			       h.len, h.caplen);
+
+			for (i = 0; i < h.caplen; i += 16) {
+				snprinth(hline, sizeof(hline),
+					 packet, h.caplen, i);
+				printf("  %s\n", hline);
+			}
+		}
+		captured_packets++;
+	}
+	exit_pcap = NULL;
+	rc = true;
+
+	fprintf(stderr, "\n%"PRIu64" packets captured\n", captured_packets);
```

Comment:
Your use of PRIu64 causes compile errors on my Ubuntu machine, because of a missing `#include <inttypes.h>`